curvefs/src/client/warmup/warmup_manager.cpp

```diff
@@ -0,0 +1,690 @@
+/*
+ *  Copyright (c) 2023 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: 2023-01-31
+ * Author: chengyi01
+ */
+
+#include "curvefs/src/client/warmup/warmup_manager.h"
+
+#include <glog/logging.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <atomic>
+#include <deque>
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "curvefs/src/client/inode_wrapper.h"
+#include "curvefs/src/client/s3/client_s3_cache_manager.h"
+#include "curvefs/src/common/s3util.h"
+#include "src/common/concurrent/concurrent.h"
+#include "src/common/string_util.h"
+
+
+namespace curvefs {
+namespace client {
+namespace warmup {
+
+using curve::common::WriteLockGuard;
+
+#define WARMUP_CHECKINTERVAL_US (1000 * 1000)
+
+void WarmupManagerS3Impl::AddWarmupFilelist(fuse_ino_t key) {
+    if (!mounted_.load(std::memory_order_acquire)) {
+        LOG(ERROR) << "not mounted";
+    }
+    // add warmup Progress
+    if (!AddWarmupProcess(key)) {
+        return;
+    }
+    WriteLockGuard lock(warmupFilelistDequeMutex_);
+    auto iter = LockedFindKeyWarmupFilelist(key);
+    if (iter == warmupFilelistDeque_.end()) {
+        std::shared_ptr<InodeWrapper> inodeWrapper;
+        CURVEFS_ERROR ret = inodeManager_->GetInode(key, inodeWrapper);
+        if (ret != CURVEFS_ERROR::OK) {
+            LOG(ERROR) << "inodeManager get inode fail, ret = " << ret
+                       << ", inodeid = " << key;
+            return;
```

Comment:
maybe need delete the value added upside？and  also do we need notice the users？

Reply:
> maybe need delete the value added upside？and also do we need notice the users？

fix

Reply:
> maybe need delete the value added upside？and also do we need notice the users？

fix